src/main/java/org/candlepin/resource/util/ConsumerCloudDataBuilder.java

```diff
@@ -0,0 +1,116 @@
+import org.candlepin.model.ConsumerCloudData;
+
+import org.xnap.commons.i18n.I18n;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import javax.inject.Inject;
+
+/**
+ * Builder for constructing {@link ConsumerCloudData} from consumer facts.
+ * <p>
+ * This class selects the appropriate {@link CloudProviderFactParser} based on the provided facts and
+ * builds a {@link ConsumerCloudData} object containing cloud-specific information.
+ * </p>
+ */
+public class ConsumerCloudDataBuilder {
+
+    private final I18n i18n;
+    private final Set<CloudProviderFactParser> parsers;
+
+    @Inject
+    public ConsumerCloudDataBuilder(I18n i18n, Set<CloudProviderFactParser> parsers) {
+        this.parsers = parsers;
+        this.i18n = i18n;
+    }
+
+    /**
+     * Builds a {@link ConsumerCloudData} object based on the facts from the given consumer.
+     *
+     * @param consumer the consumer whose facts will be used for building the cloud data.
+     * @return an {@link Optional} containing the {@link ConsumerCloudData} if a supported
+     *         cloud provider is found or an empty {@link Optional} if no supported provider is found.
+     */
+    public Optional<ConsumerCloudData> build(Consumer consumer) {
+        return buildConsumerCloudData(consumer.getFacts());
+    }
+
+    /**
+     * Builds a {@link ConsumerCloudData} object based on the facts from the given consumer DTO.
+     *
+     * @param consumer the consumer DTO whose facts will be used for building the cloud data.
+     * @return an {@link Optional} containing the {@link ConsumerCloudData} if a supported cloud
+     *         provider is found, or an empty {@link Optional} if no supported provider is found.
+     */
+    public Optional<ConsumerCloudData> build(ConsumerDTO consumer) {
+        return buildConsumerCloudData(consumer.getFacts());
+    }
+
+    /**
+     * Processes the provided facts and builds a {@link ConsumerCloudData} object if exactly one
+     * {@link CloudProviderFactParser} supports the facts.
+     * <p>
+     * If more than one parser supports the facts, a {@link BadRequestException} is thrown.
+     * If no parser supports the facts, an empty {@link Optional} is returned.
+     * </p>
+     *
+     * @param facts the map of cloud provider facts.
+     * @return an {@link Optional} containing the {@link ConsumerCloudData} or an empty {@link Optional}.
+     * @throws BadRequestException if more than one parser supports the provided facts.
+     */
+    private Optional<ConsumerCloudData> buildConsumerCloudData(Map<String, String> facts) {
+        List<CloudProviderFactParser> supportedParsers = parsers.stream()
+            .filter(parser -> parser.isSupported(facts))
+            .toList();
+
+        if (supportedParsers.size() > 1) {
+            throw new BadRequestException(this.i18n.tr("During consumer cloud creation more than one " +
+                "cloud provider facts found"));
```

Comment:
The translated text needs to be on the same line. You can use `i18n.marktr` on a separate line and use the returned string for the `BadRequestException`.